### PR TITLE
Use quantity rather than string in task resource config

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Use quantity rather than string in task resource config. Also filed https://github.com/flyteorg/flyte/issues/1403 to track making these changes more widespread

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin
 - [x] Housekeeping

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/execution_manager.go [hunk 1]
```diff
@@ -292,8 +312,8 @@ func assignResourcesIfUnset(ctx context.Context, identifier *core.Identifier,
 		}
 		resourceEntries = append(resourceEntries, cpuResource)
 	}
-	if memoryIndex < 0 && len(platformValues.Memory) > 0 {
-		memoryValue := platformValues.Memory
+	if memoryIndex < 0 && !platformValues.Memory.IsZero() {
```
Comment (kumare3) [inline pkg/manager/impl/execution_manager.go:315]: i am guessing these values cannot be nil

File: pkg/manager/impl/execution_manager.go [hunk 2]
```diff
@@ -235,26 +235,46 @@ func createTaskDefaultLimits(ctx context.Context, task *core.CompiledTask,
...
-	if len(taskResourceLimits.GPU) == 0 && len(configResourceLimits.GPU) > 0 {
+	if !configResourceLimits.CPU.IsZero() {
+		taskResourceLimits.CPU = configResourceLimits.CPU
+	} else {
+		if cpuIndex >= 0 {
+			cpuLimit = resourceRequestEntries[cpuIndex].Value
+		}
+		cpu, err := resource.ParseQuantity(cpuLimit)
+		if err != nil {
+			logger.Infof(ctx, "Failed to parse user cpu limit from task spec [%s] with err [%+v]", cpuLimit, err)
```
Comment (EngHabu) [inline pkg/manager/impl/execution_manager.go:246]: Shouldn't this fail the task?

Comment (EngHabu) [reply pkg/manager/impl/execution_manager.go:246]: I guess this should have failed the compilation...

Comment (EngHabu) [reply pkg/manager/impl/execution_manager.go:246]: Maybe make it `Errorf` and indicate this should have failed in the compiler earlier?

File: pkg/runtime/task_resource_provider.go [hunk 3]
```diff
@@ -3,26 +3,59 @@ package runtime
...
 
 var taskResourceConfig = config.MustRegisterSection(taskResourceKey, &TaskResourceSpec{})
 
 type TaskResourceSpec struct {
-	Defaults interfaces.TaskResourceSet `json:"defaults"`
-	Limits   interfaces.TaskResourceSet `json:"limits"`
+	Defaults         interfaces.DeprecatedTaskResourceSet `json:"defaults"`
+	DefaultResources interfaces.TaskResourceSet           `json:"defaultResources"`
+	Limits           interfaces.DeprecatedTaskResourceSet `json:"limits"`
+	LimitResources   interfaces.TaskResourceSet           `json:"limitResources"`
```
Comment (EngHabu) [inline pkg/runtime/task_resource_provider.go:17]: Does that mean the administrator will need to migrate configs?

Comment (EngHabu) [reply pkg/runtime/task_resource_provider.go:17]: talked offline, we don't need the migration path... we can just change the types from string to resource.Quantity....

**Review by EngHabu — COMMENTED** [review]
Looks good... just the comment about migration and we can merge this! Thank you!!

File: pkg/manager/impl/execution_manager.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go:246]: oh yeah this is why i filed https://github.com/flyteorg/flyte/issues/1403 to clean this up so we can easily enforce this at serialization/compilation

Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: upgraded to an error

**Comment by katrogan** [comment]
PTAL @EngHabu

**Review by EngHabu — APPROVED** [review]
